File: python/leetcode/0509_fibonacci_number.py
from assertion import assert_equal

class Solution:
    """
    F(0) = 0, F(1) = 1
    F(n) = F(n - 1) + F(n - 2), for n > 1.

    O(n), O(2)
    """
    def fib(self, n: int) -> int:
        dp = [0, 1]

        for i in range(2, n+1):
            dp[i % 2] = dp[0] + dp[1]

        return dp[n % 2]
    
if __name__ == '__main__':
    assert_equal(Solution().fib(2), 1)
    assert_equal(Solution().fib(3), 2)
    assert_equal(Solution().fib(4), 3)
    

File: ts/src/ts/0509_fibonacci_number.ts
import assertEqual from './assert';

/*
 */
function fib(n: number): number {
  const dp = [0, 1];

  for (let i = 2; i <= n; i++) {
    dp[i % 2] = dp.reduce((a, b) => a + b);
  }

  return dp[n % 2] as number;
};

assertEqual(fib(2), 1);
assertEqual(fib(3), 2);
assertEqual(fib(4), 3);



